### ISSUE TITLE
Scalable Side Navigation Bar Design

### ISSUE THREAD

**Issue by Expalex** [backstage/backstage#6933]
Expedia Group would like to propose side navigation designs that would be more scalable.
The conversation was [initially](https://github.com/backstage/backstage/issues/5826) started a few months ago to explore a design that could be added to the Storybook.
 
Please view the Side [Nav v1.2_Video](https://drive.google.com/file/d/1gciX9w0qw1RARMNNGfwZuEgkWotxrr5P/view?usp=sharing) created using the Figma prototype [Side Nav_v1.2](https://www.figma.com/proto/df353LDD1WiXh78M9a3jK9/Side-Navigation?page-id=229%3A567&node-id=229%3A3625&viewport=267%2C48%2C1&scaling=contain&starting-point-node-id=229%3A3625&show-proto-sidebar=1) .
Our recommendation is the more scalable option of the two designs, but would like your feedback on both.
 
To start, we are interested in exploring ideal design architecture and flows together. Contrast ratios and colors can be adjusted later. Our [current settings](https://drive.google.com/file/d/1Lur78PYKc1VCQw5VjZq-Jg9XHvjDtavc/view?usp=sharing) includes profile and other features. Profile and setting in the prototype has been placed as what might be nice to have, and hear your thoughts on the future role based access controls.

**Comment by nwpappas** [comment]
This is exciting to see, as we've been exploring similar concepts. A few comments/questions I'd enjoy hearing your finding on, from the video:

What does 'Search' actually search and how are the result intended to be displayed?

Did you explore a click-to-open on the secondary menu for Option 1, instead of retaining the mouse over? We've found the click-to-open to be less distracting to users, for exactly the reasons you mention at the start for the primary navigation automatically opening.

**Comment by Expalex** [comment]
Enabling custom add menus and submenus will really heighten the navigation’s usability.
@nwpappas  If kept simple, I expect search to bring up and autofill with menu names. To make it more advanced, menu types can be searched as well, but a lot can be done. What are you picturing?
Yes, as you said I have thought the possibility of click-to-open. That could work too. There are several reasons for the mouse over fly open of the multi category menu:
1. Minimize clicks
2. Maintain consistency with the single mouse over click to select. Of course, sub menus on the side column bar will require more.
3. Some similarity to the existing Backstage design

**Comment by Expalex** [comment]
Enabling custom add menus and submenus will really heighten the navigation’s usability.

@nwpappas
If kept simple, I expect search to bring up and autofill with menu names. To make it more advanced, menu types can be searched as well, but a lot can be done. What are you picturing?
Yes, as you said I have thought the possibility of click-to-open. That could work too. There are several reasons for the mouse over fly open of the multi category menu:
1) Minimize clicks
2) Consistency with the single mouse over click to select. Of course, sub menus on the side column bar will require more.
3) Maintain some similarity to the existing Backstage design

**Comment by Rugvip** [comment]
I do like option 1 :grin:

@katz95 for visibility as well

From a technical angle the `Sidebar` is intentionally very loosely coupled to the rest of the app. I'm thinking it would make sense to move the sidebar components out of `@backstage/core-components` into something like `@backstage/sidebar`, and with that perhaps build it out a bit to allow for some more customization.

**Comment by freben** [comment]
Are we really at a point where the component library would benefit from package splitting? :) I'd spontaneously vote for instead putting this in the same package and naming it for example `NestedSidebar` or similar.

**Comment by freben** [comment]
Hmmm. `app-components`, `plugin-components`, `common-components`? :)

**Comment by Rugvip** [comment]
@freben yeah essentially that. It's kind of what we actually have but it felt a bit overkill

**Comment by Expalex** [comment]
Sounds like option 1 is a possible design to explore further. I’ll improve the prototype and share. Is there anything else the 
Expedia Group can provide?

**Comment by freben** [comment]
@Rugvip to get experimentation going on this cool contrib - is this the time to break it out into a separate lib or can we let this start to get added in `core-components` as an alpha and then consider refactoring the structure of all of these things separately? Not sure that `@backstage/sidebar` is the right granularity for example

**Comment by KatherineZhou** [comment]
Hi!! This is so great, thanks for sharing your efforts! Showed these designs to our internal design team, and we love Option 1, as it gets users right to the subnav level and cuts out redundancy. Perhaps we should do some testing with users to see if they find hover annoying or not. A few questions:

1. Is that nested Search in the expandable subnav only for the entities within that subnav? Wondering if that's a bit confusing (as in, users might think it's global).
2. The dropdown menu inside the hover — does that mean when I click on “Misc” it won’t take me to a page? 
3. Attached is some suggestions from my team
<img width="1112" alt="sidenav" src="https://user-images.githubusercontent.com/16808998/131254318-bae90ccd-6283-49b7-a7e1-c371cb417b40.png">

**Comment by Expalex** [comment]
@KatherineZhou  Thank you so much for your suggestions.
 1)
Search would only be for within the subnav, which in this example is “Catalog”. It’s scope can be clarified by stating, “Search Catalog” or something like that. Width of the subnav would be customizable, allowing it to expand further. 
It’d match the length of the longest menu label. In this example, it’d be “Lorem Ipsum 2”. Search would auto match and populate with available menu labels as a user enters search words. On this screenshot, I made the subnav and search a little wider than what’s shown on the video. Of the three search sizes, 1 or 2 should be sufficient. We’ll do user research for optimal search scenarios and design.
<img width="1199" alt="Search_sizes" src="https://user-images.githubusercontent.com/80132060/131417417-42113be8-613c-4326-a55d-fbf031abc6a7.png">
2)
I am glad you brought this up. It seems like a good idea to give both options - for sub category section header such as “Misc” to act 
  A) as a label and only those below will be enabled to change pages
  B) to change pages also

Additional insights and suggestions would be appreciated.

**Comment by timbonicus** [comment]
Piling on with some feedback. First off, big kudos for the video to walk through these options and the rationale! That was awesome to see. 🙇  I think the highlight on hover, and pop-out submenu of option 1 are really awesome. I also like having the expand toggle.

I wondered, like Kat did, whether I could click on the header of a sub-menu. For example, how do I get to a list of _all things_ in the catalog rather than a view filtered by Services, etc? As presented, I'd need to be able to click on the base catalog icon. This has the potential to be confusing if someone tries clicking to open the menu instead of hovering. Another option I like better, if this _wasn't_ clickable, is to add an `All` option to the sub-menu to handle those cases (when appropriate). This also keeps a consistent UX when an `All` option doesn't make sense for a particular menu.

I also agree with Kat that the search is confusing. Even with a `Search Catalog` label, it appears that it will search through the **actual contents** of the software catalog -- not search through _menu items_. It's right next to the other search menu. I would personally argue that if your menu scales enough to need a _search bar_, maybe you're escaping the limits of IA that a menu can solve and another approach needs to be considered (global vs in-page nav, for example). 😅

**Comment by freben** [comment]
For starters, this is totally awesome and I'm really impressed and happy about the initiative and design!

That being said, agreed on the topic of search. I think the word "search" should be strictly reserved for the global search experience, and not used for other functions. Three different paths I could envision:

- Keep the proposed function but name it "Filter" or similar, just like one does in tables that have a quick filter function at the top. I am still a bit doubtful, as @timbonicus mentions, that this has significant impact in a little sidebar, and might imply a bit of a design/nav smell if they get so large that such a function feels necessary
- Do something like the filter on system preferences in operating systems, where you can filter an entire huge set of tools down to the ones you need (i.e. have the filter at the root rater than in each column)
- Keep "Search catalog" but make it a button that opens the actual search experience, pre-filtered to only show catalog stuff. THAT search experience should have a great "empty state" (no text entered) behavior, showing the most popular filters, your recently used, etc, until any typing starts to happen

But in essence, I think that maybe the search thing could perhaps be skipped in the initial implementation? And then think about that one some more and see what it should turn out to look like.

**Comment by Expalex** [comment]
Wonderful feedbacks and suggestions. I love them all. They have been discussed with our team and the following might be good to do.
   1. Search Menu:   skip in the initial implementation.
   2. Submenu Open Method:   mouseover slide open design as shown on the video and Figma mockup would allow for a consistent and single click experience. This would be especially helpful on menu scenarios like Catalog, which has most used apps such as Services, APIs, and Tools. Minimizing clicks this way will provide better user experience. To clarify, the arrow next to the icon indicates that there is a submenu. Users who prefer click to open, can enable that feature in the settings.
   3. Multi-Category Subjects:  such as Misc should be allowed to click to show a page by default. However, this can be turned off to allow only the subcategory menus to be clickable, by simple flag changes in the code.

**Comment by freben** [comment]
Oh on the topic of hover: Do consider touch-only devices. They will want to use touch events (like clicks) since hover isn't really a thing there. You'd probably want to make both work (open on hover OR touch)

**Comment by Expalex** [comment]
@freben Good point. I am still relatively new to Backstage, and have been designing around common desktop usage. I’ll discuss further with our engineering team for design that provides optimal experience across devices.

**Comment by SuganJoe** [comment]
Is everyone is good with @Expalex design along with improvements in his recent comment?

If yes, from the Expedia group we would like to get the issue assigned to us and start implementing the feature.

**Comment by KatherineZhou** [comment]
Hi!! Thanks for getting back! Agreed with skipping the search menu in the initial implementation.

Also agreed with @freben regarding hover/touch.

Regarding this:

> Multi-Category Subjects: such as Misc should be allowed to click to show a page by default. However, this can be turned off to allow only the subcategory menus to be clickable, by simple flag changes in the code.

If I'm understanding you correctly, I'm wondering if that might present some confusion to users who might not know what to expect. They might learn to not click anything or they might get frustrated if they expect something to be clickable when it's not. I'm personally a fan of making them clickable.

**Comment by freben** [comment]
From the maintainers point of view we're happy you're taking this on and look forward to reviewing eventually. Assigned! Hook us up on discord or continue the discussion here, if you have any questions in the mean time.
